extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java

```diff
@@ -273,4 +190,105 @@ public void setSecurityEventObserved(boolean isSecurityEventObserved) {
+                                        }
+
+                                    });
+                        } else {
+                            return Uni.createFrom().item(new OidcProvider(client, oidcConfig, null));
+                        }
+                    }
+                }).await().indefinitely();
+    }
+
+    protected static Uni<OidcProviderClient> createOidcClientUni(OidcTenantConfig oidcConfig,
+            TlsConfig tlsConfig, Vertx vertx) {
+
+        OidcCommonUtils.verifyCommonConfiguration(oidcConfig);
+
+        String authServerUriString = OidcCommonUtils.getAuthServerUrl(oidcConfig);
+
+        WebClientOptions options = new WebClientOptions();
+
+        URI authServerUri = URI.create(authServerUriString); // create uri for parse exception
+        OidcCommonUtils.setHttpClientOptions(oidcConfig, tlsConfig, options);
+
+        WebClient client = WebClient.create(new io.vertx.mutiny.core.Vertx(vertx), options);
+
+        Uni<OidcConfigurationMetadata> metadataUni = null;
+        if (!oidcConfig.discoveryEnabled) {
+            String tokenUri = OidcCommonUtils.getOidcEndpointUrl(authServerUri.toString(), oidcConfig.tokenPath);
+            String introspectionUri = OidcCommonUtils.getOidcEndpointUrl(authServerUri.toString(),
+                    oidcConfig.introspectionPath);
+            String authorizationUri = OidcCommonUtils.getOidcEndpointUrl(authServerUri.toString(),
+                    oidcConfig.authorizationPath);
+            String jwksUri = OidcCommonUtils.getOidcEndpointUrl(authServerUri.toString(), oidcConfig.jwksPath);
+            String userInfoUri = OidcCommonUtils.getOidcEndpointUrl(authServerUri.toString(), oidcConfig.userInfoPath);
+            String endSessionUri = OidcCommonUtils.getOidcEndpointUrl(authServerUri.toString(), oidcConfig.endSessionPath);
+            metadataUni = Uni.createFrom().item(new OidcConfigurationMetadata(tokenUri,
+                    introspectionUri, authorizationUri, jwksUri, userInfoUri, endSessionUri,
+                    oidcConfig.token.issuer.orElse(null)));
+        } else {
+            final long connectionRetryCount = OidcCommonUtils.getConnectionRetryCount(oidcConfig);
+            final long expireInDelay = OidcCommonUtils.getConnectionDelayInMillis(oidcConfig);
+            if (connectionRetryCount > 1) {
+                LOG.infof("Connecting to IDP for up to %d times every 2 seconds", connectionRetryCount);
+            }
+            metadataUni = discoverMetadata(client, authServerUri.toString(), oidcConfig).onFailure(ConnectException.class)
+                    .retry()
+                    .withBackOff(CONNECTION_BACKOFF_DURATION, CONNECTION_BACKOFF_DURATION)
+                    .expireIn(expireInDelay);
+        }
+        return metadataUni.onItemOrFailure()
+                .transformToUni(new BiFunction<OidcConfigurationMetadata, Throwable, Uni<? extends OidcProviderClient>>() {
```

Comment:
This is synchronous, use `transform` instead.

Reply:
Hi @cescoffier, yeah, I've already used `transform` alone a few times, but the reason I've used `transformToUni` here is that the exception at the start up is easier to follow. 
If we have `transform`:
```
Caused by: io.smallrye.mutiny.CompositeException: Multiple exceptions caught:
	[Exception 0] java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178544958/1614178542736 expiration
	[Exception 1] io.quarkus.oidc.OIDCException: OIDC server is not available at the 'https://localhost:8543/auth/realms/quarkus' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'
	at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureMap$UniOnItemOrFailureMapProcessor.onFailure(UniOnItemOrFailureMap.java:59)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.onError(UniCreateFromPublisher.java:79)
...
io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onFailure(UniToMultiPublisher.java:86)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:44)
	at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:36)
	at io.vertx.mutiny.ext.web.client.HttpRequest$9.handle(HttpRequest.java:619)
        ....
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: io.quarkus.oidc.OIDCException: OIDC server is not available at the 'https://localhost:8543/auth/realms/quarkus' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'
		at io.quarkus.oidc.runtime.OidcRecorder.toOidcException(OidcRecorder.java:196)
		at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:265)
		at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:260)
		at io.smallrye.context.impl.wrappers.SlowContextualBiFunction.apply(SlowContextualBiFunction.java:21)
		at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureMap$UniOnItemOrFailureMapProcessor.onFailure(UniOnItemOrFailureMap.java:54)
		... 97 more
	Caused by: java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178544958/1614178542736 expiration
		at io.smallrye.mutiny.helpers.ExponentialBackoff.lambda$randomExponentialBackoffFunctionExpireAt$2(ExponentialBackoff.java:113)
		at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
		at io.smallrye.mutiny.groups.MultiOnItem.lambda$transformToUni$7(MultiOnItem.java:354)
		at io.smallrye.mutiny.operators.multi.MultiFlatMapOp$FlatMapMainSubscriber.onItem(MultiFlatMapOp.java:176)
		... 75 more
	Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8543
	Caused by: java.net.ConnectException: Connection refused
		at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
		...
Caused by: [CIRCULAR REFERENCE: java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178544958/1614178542736 expiration]
``` 

with `transformToUni`:
```
Caused by: io.quarkus.oidc.OIDCException: OIDC server is not available at the 'https://localhost:8543/auth/realms/quarkus' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'
	at io.quarkus.oidc.runtime.OidcRecorder.toOidcException(OidcRecorder.java:196)
	at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:265)
	at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:260)
	at io.smallrye.context.impl.wrappers.SlowContextualBiFunction.apply(SlowContextualBiFunction.java:21)
	at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap$UniOnItemOrFailureFlatMapProcessor.performInnerSubscription(UniOnItemOrFailureFlatMap.java:86)
...
 io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onFailure(UniToMultiPublisher.java:86)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:44)
	at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:36)
	at io.vertx.mutiny.ext.web.client.HttpRequest$9.handle(HttpRequest.java:619)
	...
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178770278/1614178768070 expiration
	at io.smallrye.mutiny.helpers.ExponentialBackoff.lambda$randomExponentialBackoffFunctionExpireAt$2(ExponentialBackoff.java:113)
	at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
	at io.smallrye.mutiny.groups.MultiOnItem.lambda$transformToUni$7(MultiOnItem.java:354)
	at io.smallrye.mutiny.operators.multi.MultiFlatMapOp$FlatMapMainSubscriber.onItem(MultiFlatMapOp.java:176)
	... 75 more
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8543
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	...
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

There are fewer nested traces in the latter case and no `CIRCULAR_REFERENCE`. This is only reason I've kept it as `transformToUni` for now

Reply:
I don't really mind much, I thought it was a bit easier to get the actual error in the latter case...

Reply:
You mean it's creating a CompositeException with the failure and the thrown exception. 
So, yes, using Uni avoids this.

@jponge, I got that comment twice in two days. In onItemOfFailure, should we really build a composite exception?

Reply:
@cescoffier I created a quick issue, we'll investigate this https://github.com/smallrye/smallrye-mutiny/issues/488

Reply:
thanks @jponge

Reply:
@cescoffier @jponge OK, thanks :-)